File: src/test/resources/dev/hilla/generator/basic/CustomTypesEndpoint.ts
import type { MyLibraryType } from '../../../../com/example/library/MyLibraryType.js';
import client from '../../../../connect-client.default.js';
import type { MyCustomType } from './CustomTypesEndpoint/MyCustomType.js';
import type { MyOtherType } from './entities/MyOtherType.js';
import type { EndpointRequestInit } from '@hilla/frontend';

async function externalLibraryType(argument: MyLibraryType, init?: EndpointRequestInit): Promise<MyLibraryType> {
    return client.call('dev.hilla.generator.basic.CustomTypesEndpoint', 'externalLibraryType', { argument }, init);
}

async function fromArrayOfCustomsToCustom(argument: MyOtherType[], init?: EndpointRequestInit): Promise<MyOtherType> {
    return client.call('dev.hilla.generator.basic.CustomTypesEndpoint', 'fromArrayOfCustomsToCustom', { argument }, init);
}

async function fromCustomToListOfCustoms(argument: MyCustomType, init?: EndpointRequestInit): Promise<MyCustomType[]> {
    return client.call('dev.hilla.generator.basic.CustomTypesEndpoint', 'fromCustomToListOfCustoms', { argument }, init);
}

const CustomTypesEndpoint = {
    externalLibraryType,
    fromArrayOfCustomsToCustom,
    fromCustomToListOfCustoms,
};

export default CustomTypesEndpoint;

File: generator/src/test/java/dev/hilla/generator/basic/CustomTypesEndpoint.java
package dev.hilla.generator.basic;

import com.example.library.MyLibraryType;
import dev.hilla.generator.basic.entities.MyOtherType;
import dev.hilla.parser.annotations.Endpoint;
import java.util.List;

@Endpoint
public class CustomTypesEndpoint {
  public List<MyCustomType> fromCustomToListOfCustoms(MyCustomType argument) {
    return List.of(argument);
  }

  public MyOtherType fromArrayOfCustomsToCustom(MyOtherType[] argument) {
    return argument[0];
  }

  public MyLibraryType externalLibraryType(MyLibraryType argument) {
    return argument;
  }

  public static class MyCustomType {
    private String field;

    public String getField() {
      return field;
    }

    public void setField(String field) {
      this.field = field;
    }
  }
}
